packages/machine/src/middleware/middleware.ts

```diff
@@ -61,153 +33,39 @@ export class Middleware {
         }
       }
     ],
-    [Opcode.STATE_TRANSITION_PROPOSE]: [
-      {
-        scope: Opcode.STATE_TRANSITION_PROPOSE,
-        method: (message, next, context) => {
-          const proposal = StateTransition.propose(
-            message,
-            next,
-            context,
-            this.node
-          );
-          context.intermediateResults.proposedStateTransition = proposal;
-        }
-      }
-    ]
+    [Opcode.STATE_TRANSITION_PROPOSE]: []
   };
 
-  constructor(readonly node: Node) {
-    this.initializeMiddlewares();
-  }
-
-  private initializeMiddlewares() {
-    this.add(Opcode.KEY_GENERATE, KeyGenerator.generate);
-    this.add(Opcode.OP_SIGN_VALIDATE, SignatureValidator.validate);
-  }
+  constructor(readonly node: Node) {}
 
   public add(scope: Opcode, method: InstructionMiddlewareCallback) {
     this.middlewares[scope].push({ scope, method });
   }
 
   public async run(msg: InternalMessage, context: Context) {
-    let counter = 0;
     const middlewares = this.middlewares;
     const opCode = msg.opCode;
 
-    this.executeMiddlewaresRegisteredOnAll(msg, context);
-
-    async function callback() {
-      if (counter === middlewares[opCode].length - 1) {
-        return null;
-      }
-      // This is hacky, prevents next from being called more than once
-      counter += 1;
-      const middleware = middlewares[opCode][counter];
-      if (opCode === Opcode.ALL || middleware.scope === opCode) {
-        return middleware.method(msg, callback, context);
-      }
-      return callback();
+    for (const middleware of middlewares[opCode]) {
+      await middleware.method(msg, () => {}, context);
```

Comment:
The callback argument is not correctly implemented here

Reply:
I am choosing not to support this argument anymore

Reply:
I think it should be, as we previously discussed privately.

Reply:
can you summarize your argument here?

I would summarize mine as two points

1. I do not see any use-cases for supporting custom hooks apart from the existing ones implemented - messaging and signing
2. a promise-based interface between the existing custom hooks and the machine is sufficient

Reply:
(will discuss on wednesday, keeping `next` for now)